pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -288,6 +288,19 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun) err
 		return nil
 	}
 
+	// build DAG with a list of final tasks, this DAG is used later to identify
+	// if a task in PipelineRunState is final task or not
+	// the finally section is optional and might not exist
+	// dfinally holds an empty Graph in the absence of finally clause
+	dfinally, err := dag.Build(v1beta1.PipelineTaskList(pipelineSpec.Finally))
+	if err != nil {
+		// This Run has failed, so we need to mark it as failed and stop reconciling it
+		pr.Status.MarkFailed(ReasonInvalidGraph,
+			"PipelineRun %s's Pipeline DAG is invalid for finally clause: %s",
+			pr.Namespace, pr.Name, err)
+		return nil
+	}
```

Comment:
Pardon me my naive question, but any reason to build a dag *for* `Finally` tasks ? :thinking: 
As far as I understood, the tasks under `finally` are not a dag, just a list to execute (in parallel right ?)

Reply:
i think it's being used as a grouping for Tasks - but looking at the code, the finally tasks probably dont _need_ a DAG object

on the other hand, i bet you we are a short feature request away from letting finally tasks be a DAG so im not sure this is worth changing right now

(and i think technically you could say the finally tasks _are_ a dag, its just that each root doesn't have any .... edges? not 100% sure im using the right terminology XD)

if we DO want to change this, i suggest we do it in a follow up PR - i would be hesitant tho if we think it's likely we'll let finally tasks have more complicated graphs, and im betting we will eventually

Reply:
thanks @vdemeester, its never too late 🤪

Echoing what @bobcatfish explained ⬆️ 

Building a dag for `finally` allows us to reuse existing code path instead of rewriting it for different data structure or casting existing code to a `list`. 

Many existing use cases will be simplified by following this new paradigm and expecting immediate feature request to make it true DAG.

If needed, I can create a follow up PR to rewrite this and have finally tasks as a list instead of DAG.